src/client/producer.rs

```diff
@@ -323,64 +326,246 @@ impl ProducerClient for PartitionClient {
+
+                // Perform an immediate flush of the buffer in the background,
+                // returning without waiting for the flush to complete.
+                //
+                // This call sets a new BatchBuilder while the flush of the old
+                // instance happens in the background, enabling the caller's
+                // Mutex to be dropped, so that further produce() calls can
+                // start aggregating into a new batch of writes.
+                //
+                // As a side effect, this invalidates any callers performing a
+                // linger wait + flush, preventing them from flushing this new
+                // batch.
+                self.flush(None)?;
+
+                match self.batch_builder.as_mut().unwrap().try_push(data)? {
+                    TryPush::Aggregated(handle) => handle,
+                    TryPush::NoCapacity(_) => {
+                        error!(client=?self.client, "record too large for aggregator");
+                        return Err(Error::TooLarge);
+                    }
+                }
+            }
+        };
+
+        // If this is the first writer to this batch, has_linger_waiter will be
+        // false, indicating this writer should wait for the linger time before
+        // trying to flush this batch.
+        if self.has_linger_waiter {
+            // There is an existing caller handling the linger timeout.
+            return Ok(CallerRole::JustWait(handle));
+        }
+
+        // This caller is the first writer to this batch, and it should wait for
+        // the linger time before flushing THIS batch.
+        //
+        // While the writer is waiting to flush THIS batch, it is possible
+        // another call this function results in a NoCapacity being returned by
+        // the aggregator above, which would flush THIS batch.
+        //
+        // When this happens, the writer waiting for the linger time should NOT
+        // flush the newly created batch, so it uses the flush clock as a token
+        // to avoid flushing the wrong batch when calling flush().
+        self.has_linger_waiter = true;
+        Ok(CallerRole::Linger {
+            handle,
+            flush_token: self.flush_clock,
+        })
+    }
+
+    /// Asynchronously write this batch of writes to Kafka, flushing the
+    /// underlying [`Aggregator`].
+    ///
+    /// If the caller provides a `flusher_token`, the batch flush is conditional
+    /// on the token matching. If the token does not match, the batch the caller
+    /// is attempting to flush has already been flushed, and this call is a NOP.
+    fn flush(&mut self, flusher_token: Option<usize>) -> Result<()> {
+        // If this caller is is intending to conditionally flush a specific
+        // batch, verify this BatchBuilder is the batch it is indenting to
+        // flush.
+        if let Some(token) = flusher_token {
+            if token != self.flush_clock {
```

Comment:
Can this only happen when there are manual flushes involved or are there other cases as well?

Reply:
This block of code intentionally doesn't care why the tokens don't match!

But to answer properly, this can happen two ways:

* A manual call to [`BatchProducer::flush()`] causes the buffer to be flushed early. The caller doing the linger either:
    * [Observes the result of the flush early, before the linger expires][early], or
    * The linger fires after the manual `flush()`, but before the result is broadcast. In this case, [this conditional flush][after] becomes a NOP.

* At least two writes to an aggregator are made; the first caller starts lingering, and the second caller [observes a full aggregator][full] and flushes the batch before the linger expires. The subsequent flush attempt by the linger is a NOP (which may also take the same early flush path as above).

[full]:https://github.com/influxdata/rskafka/blob/82862df069978a6fcca52567ddf16bda85d4a8cc/src/client/producer.rs#L430-L457
[after]:https://github.com/influxdata/rskafka/blob/82862df069978a6fcca52567ddf16bda85d4a8cc/src/client/producer.rs#L654
[early]:https://github.com/influxdata/rskafka/blob/82862df069978a6fcca52567ddf16bda85d4a8cc/src/client/producer.rs#L663
[`BatchProducer::flush()`]: https://github.com/influxdata/rskafka/blob/82862df069978a6fcca52567ddf16bda85d4a8cc/src/client/producer.rs#L681